### ISSUE TITLE
Is there a small model trainer?

### ISSUE THREAD

**Issue by win10ogod** [state-spaces/mamba#24]
Is there a small model trainer, such as llama2.c?

**Comment by albertfgu** [comment]
We do not have one currently. Hopefully the community will help out!

**Comment by pjsample** [comment]
I haven't tried it but this repo appears to implement a trainer.
https://github.com/havenhq/mamba-chat/tree/main

**Comment by masc-it** [comment]
A simple script to pretrain mamba using the Conditional Language Modeling (CLM) task, extrapolated from mamba-chat and adapted for CLM:

```python
from transformers import Trainer
import torch
import os
import torch
import argparse
from datasets import load_from_disk
from mamba_ssm.models.mixer_seq_simple import MambaLMHeadModel
from transformers import AutoTokenizer, TrainingArguments, TrainerCallback
from trainer.mamba_trainer import MambaTrainer
from transformers import DataCollatorForLanguageModeling

class MambaTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False):
        input_ids = inputs.pop("input_ids")
        lm_logits = model(input_ids).logits

        labels = input_ids.to(lm_logits.device)
        shift_logits = lm_logits[:, :-1, :].contiguous()
        labels = labels[:, 1:].contiguous()

        loss_fct = torch.nn.CrossEntropyLoss()
        lm_loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), labels.view(-1))

        return lm_loss

    def save_model(self, output_dir, _internal_call):
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
            
        torch.save(self.model.state_dict(), f"{output_dir}/pytorch_model.bin")
        self.tokenizer.save_pretrained(output_dir)

def train(args):
    model = MambaLMHeadModel.from_pretrained(args.model, dtype=torch.bfloat16, device="cuda")

    tokenizer = AutoTokenizer.from_pretrained(args.tokenizer)
    tokenizer.eos_token = "<|endoftext|>"
    tokenizer.pad_token = tokenizer.eos_token
    data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False) # 

    ds_tokenized = load_from_disk("your_pretokenized_ds_mamba")

    trainer = MambaTrainer(
        model=model,
        train_dataset=ds_tokenized["train"],
        tokenizer=tokenizer,
        args=TrainingArguments(
            learning_rate=args.learning_rate,
            num_train_epochs=args.num_epochs,
            per_device_train_batch_size=args.batch_size,
            gradient_accumulation_steps=args.gradient_accumulation_steps,
            dataloader_num_workers=2,
            optim=args.optim,
            output_dir="out",
            logging_steps=50,
            weight_decay=1e-2,
            evaluation_strategy="epoch",
            save_strategy="epoch"
        ),
        
        data_collator=data_collator
    )

    trainer.train()
```
I'm having issues with the eval step, so I've omitted it.

**Comment by Eupham** [comment]
Sorry I'm trying to figure out where we get trainer.mamba_trainer from.

**Comment by masc-it** [comment]
@Eupham snipped fixed, it was just some old import.

**Comment by Eupham** [comment]
Is there any chance of a notebook showing how to train with this?  I'm doing something wrong in my attempts.

**Comment by Eupham** [comment]
Nevermind.  This works for me.  Adapted from [here](https://huggingface.co/Q-bert/Mamba-1B)
`import torch
from transformers import AutoTokenizer, TrainingArguments, Trainer
from datasets import load_dataset, load_metric
import numpy as np
from mamba_ssm.models.mixer_seq_simple import MambaLMHeadModel

model_name = "state-spaces/mamba-130m"  # Make sure this model is compatible with MambaLMHeadModel
CHAT_TEMPLATE_ID = "HuggingFaceH4/zephyr-7b-beta"
device = "cuda:0" if torch.cuda.is_available() else "cpu"

tokenizer = AutoTokenizer.from_pretrained("HuggingFaceH4/zephyr-7b-beta")
tokenizer.eos_token = ""
tokenizer.pad_token = tokenizer.eos_token
tokenizer.chat_template = AutoTokenizer.from_pretrained(CHAT_TEMPLATE_ID).chat_template

model = MambaLMHeadModel.from_pretrained(model_name)  # Changed to MambaLMHeadModel
model.to(device)

dataset = load_dataset('databricks/databricks-dolly-15k')

def tokenize_function(examples):
    concatenated_texts = [instr + " " + ctxt + " " + resp
                          for instr, ctxt, resp in zip(examples['instruction'],
                                                       examples['context'],
                                                       examples['response'])]
    return tokenizer(concatenated_texts, truncation=True, padding='max_length', max_length=500)

tokenized_datasets = dataset.map(tokenize_function, batched=True, num_proc=12)

accuracy_metric = load_metric("accuracy")

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    return accuracy_metric.compute(predictions=predictions, references=labels)

training_args = TrainingArguments(
    output_dir="./model_output",
    num_train_epochs=1,
    per_device_train_batch_size=1,
    save_steps=10_000,
    save_total_limit=2,
    logging_dir="./logs",
    logging_steps=500,
    do_train=True,
    fp16=False
)

class MambaTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False):
        input_ids = inputs.pop("input_ids")
        lm_logits = model(input_ids).logits

        labels = input_ids.to(lm_logits.device)
        shift_logits = lm_logits[:, :-1, :].contiguous()
        labels = labels[:, 1:].contiguous()

        loss_fct = torch.nn.CrossEntropyLoss()
        lm_loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), labels.view(-1))

        return lm_loss

trainer = MambaTrainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets["train"],
    compute_metrics=compute_metrics
)

trainer.train()
`

**Comment by zabealbe** [comment]
The suggested MambaTrainer code trains the model to learn the padding tokens as well.

Why? Huggingface collators such as `DataCollatorForLanguageModeling` sets the padding tokens to the integer value of -100, which is ignored by `torch.nn.CrossEntropyLoss()`, but it does so only for the `labels` not for the `input_ids`.

To prevent the model being trained on predicting padding tokens I found it's sufficient to change the code below
```python
class MambaTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False):
        input_ids = inputs.pop("input_ids")
        lm_logits = model(input_ids).logits

        labels = input_ids.to(lm_logits.device)
        shift_logits = lm_logits[:, :-1, :].contiguous()
        labels = labels[:, 1:].contiguous()

        loss_fct = torch.nn.CrossEntropyLoss()
        lm_loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), labels.view(-1))

        return lm_loss
```
To
```python
class MambaTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False):
        input_ids = inputs.pop("input_ids")
        labels = inputs.pop("labels")
        lm_logits = model(input_ids).logits

        labels = labels.to(lm_logits.device)
        shift_logits = lm_logits[:, :-1, :].contiguous()
        labels = labels[:, 1:].contiguous()

        lm_loss = torch.nn.CrossEntropyLoss()(
            shift_logits.view(-1, shift_logits.size(-1)), labels.view(-1)
        )

        return (lm_loss, lm_logits) if return_outputs else lm_loss
```
